### PR TITLE
Pprintast: add parentheses around `fun (type t) -> ...` and in `(lazy ...)` patterns

### PR DESCRIPTION
**Pull request by nojb**

### CODE REVIEW

**Comment by gasche** [comment]
I suppose you have examples that show that those parentheses are necessary for correct reparsing?

**Comment by nojb** [comment]
Yes: `f (fun (type t) -> ...)` will be pretty printed as `f fun (type t) -> ...` and `(lazy (A ...))` is printed as `(lazy A ...)`. Do you think a test is in order? (not sure adding individual counterexamples is the most efficient way of testing `Pprintast`.)

**Comment by gasche** [comment]
If I remember correctly, some people ( @Drup ? ) have tested the fact that adding a printing-reparsing step to the compiler preserves compilability of the compiler distribution. In that spirit, having these patterns occur anywhere in the codebase or testsuite should be enough to (eventually, as in "eventual consistency") detect  the issue -- typically in `exotic_syntax`. But yes, I would think that individual failures warrant individual regressions, which can catch issues later in a refactoring -- or at least increase confidence in a result.

I suppose that there is some logic to detect `Pexp_newtype` adjacent to other `Pexp_newtype` or `Pexp_fun`, and print them as a n-ary abstraction `fun x (type t) ..`?

I think that it is a bug to have placed `Pexp_newtype` in `simple_expr`, which should be reserved to fully-enclosed forms. Instead of enforcing parentheses, it should probably be handled like `Pexp_fun`: at the `expression` level, with the weird `when ctxt.semi` test as well. Would you mind trying this?

**Comment by nojb** [comment]
Yes, I will give it a try.

**Comment by nojb** [comment]
@gasche Moved the treatment of `Pexp_newtype` to be the same as `Pexp_fun` and added a test.

File: testsuite/tests/parsetree/source.ml [hunk 1]
```diff
@@ -7364,3 +7364,11 @@ let x = ` (* wait for it *) Bar
 type (+' a, -' a', ' a'b', 'ab', ' abcd', ' (* ! *) x) t =
   ' a * ' a' * ' a'b' * 'ab' * ' abcd' * ' (* !! *) x
   as ' a'
+
+(* GPR#2190 *)
```
Comment (gasche) [inline testsuite/tests/parsetree/source.ml]: You could drop the `GPR` prefix now.

**Review by gasche — APPROVED** [review]
This looks good to me, thanks. See a minor comment inline, and please feel free to merge once the CI agrees.

File: testsuite/tests/parsetree/source.ml (hunk 1 above)
Comment (nojb) [reply testsuite/tests/parsetree/source.ml]: Yes, done.

**Pull request merged**